### ISSUE TITLE
Dynamic imports with systemjs

### ISSUE THREAD

**Issue by charlag** [PepsRyuu/nollup#110]
Hi, thanks for such a cool project!
I am trying to set up the dev server and I'm running into issues with dynamic imports.

Here's one example.

```js
lang.init().then(() => {
        //...
	const workerPromise = import("./api/main/WorkerClient.js")
		.then(module => module.worker)
})
```

is translated to

```js
  const workerPromise = require.dynamic(\'/home/ivk/dev/repositories/tutanota-3/src/api/main/WorkerClient.js\').then(module => module.worker);\n\
  workerPromise.then(() => {\n\
    require.dynamic(\'/home/ivk/dev/repositories/tutanota-3/src/gui/InfoMessageHandler.js\');\n\
  });\n\
```

which is correct

but

```js
class Header {
	constructror() {
		import(`../../api/main/WorkerClient.js`)
			.then(workerClientModule => {
				(workerClientModule.worker: WorkerClient).wsConnection().map(state => {
					this._wsState = state
					m.redraw()
				})
			})
	}
}
```

is translated like

```js
import(`../../api/main/WorkerClient.js`).then(workerClientModule => {\n\
      workerClientModule.worker.wsConnection().map(state => {\n\
        this._wsState = state;\n\
        m.redraw();\n\
      });\n\
    });\n\
```

(note that import is left as-is)

and also 
```js
class WindowFacade {
	constructor() {
		import(`../api/main/WorkerClient.js`)
			.then(module => {
				// load async to reduce size of boot bundle
				this._worker = module.worker
				return nativeApp.initialized()
			}).then(() => this.addPageInBackgroundListener())
	}
}
```

is translated like 
```js
    import(`../api/main/WorkerClient.js`).then(module => {\n\
      // load async to reduce size of boot bundle\n\
      this._worker = module.worker;\n\
      return nativeApp.initialized();\n\
    }).then(() => this.addPageInBackgroundListener());\n\
```

Here's  the config which I use with dev server:

```js
const SystemConfig = require('./SystemConfig.js')
const {babel} = require("@rollup/plugin-babel")
const commonjs = require("@rollup/plugin-commonjs")
const path = require("path")

function resolveLibs(baseDir = ".") {
	return {
		name: "resolve-libs",
		resolveId(source) {
			const resolved = SystemConfig.dependencyMap[source]
			return resolved && path.join(baseDir, resolved)
		}
	}
}

function rollupDebugPlugins(baseDir) {
	return [
		babel({
			plugins: [
				// Using Flow plugin and not preset to run before class-properties and avoid generating strange property code
				"@babel/plugin-transform-flow-strip-types",
				"@babel/plugin-proposal-class-properties",
				"@babel/plugin-syntax-dynamic-import"
			]
		}),
		resolveLibs(baseDir),
		commonjs({
			exclude: ["src/**"],
			ignore: ["util"]
		}),
	]
}

module.exports = {
	input: ["src/app.js", "src/api/worker/WorkerImpl.js"],
	plugins: rollupDebugPlugins(""),
	output: {format: "system", sourcemap: "inline", dir: "build"},
}
```

I can't possibly think of the reason why some imports are not handled and I hope it's not a bug but me misusing something. Any help is appreciated!

**Comment by charlag** [comment]
Ooops, I think the problem is that they are template literals, I will close it if it's the case, sorry for the spam!

**Comment by PepsRyuu** [comment]
Yep, the problem is the template literals, strings are expected. Reason being is because template literals can contain expressions which cannot be statically evaluated. https://github.com/PepsRyuu/nollup/blob/master/lib/impl/ImportExportResolver.js#L151

Could probably add support for basic string template literals, but would need to check the behaviour of Rollup. Don't want to be inconsistent with it. :)

No problem at all! :D

**Comment by charlag** [comment]
I think that there should be a log/allowlist for unersolved imports, I remember rollup was doing this before

**Comment by PepsRyuu** [comment]
Running some tests, looks like Rollup does allow basic template literals, so I can look at adding support for that. :)

**Comment by charlag** [comment]
Thanks, this would be nice but probably the real problem for me will be really dynamic imports, I'll check when I'll get there

**Comment by PepsRyuu** [comment]
https://rollupjs.org/guide/en/#resolvedynamicimport This hook is supported. :)

**Comment by charlag** [comment]
I run into another problem with dynamic imports where import silently fails like this:

```js
                localRequire.dynamic = function (file) {
                    return new Promise(function (resolve) {
                        var relative_file = __nollup__import__map__[file];
                        var resolved_file = getResolvedPath('.', relative_file);

                        var cb = () => {
                            let id = chunks[resolved_file];
                            if (instances[number].dynamicDependencies.indexOf(id) === -1) {
                                instances[number].dynamicDependencies.push(id);
                            }

                            resolve(_require(module, id));
                        };

                        if (chunks[resolved_file]) {
                            cb();
                        } else {
                            
                            
                        }
                    });
```
nothing in the `else` case. Is it intended?

**Comment by charlag** [comment]
```js
                        if (chunks[resolved_file]) {
                            cb();
                        } else {
                            ${context.output.format === 'es'? (`
                                return import(relative_file).then(cb);
                            `) : ''}
                            ${context.output.format === 'cjs'? (`
                                return Promise.resolve(require(relative_file)).then(cb);
                            `) : ''}
                        }
```

looks like nothing in systemjs case :thinking:

**Comment by charlag** [comment]
Hm, so `__nollup_dynamic_require_callback` is called after dynamic module is loaded. But it cannot be loaded until this callback is executed :thinking:

**Comment by PepsRyuu** [comment]
SystemJS isn't a supported format at the moment. I don't know much about it, would need to look into it in more detail. Should be feasible to support, just wasn't a priority, the focus has been on common features. 

Not sure I understand the issue with ```__nollup_dynamic_require_callback```... :(

**Comment by charlag** [comment]
Oh, that's good to know. Maybe we can even add support, we will see.
So far adding case:
```js

                            ${context.output.format === "system" ? (
				'return System.import(relative_file).then(cb)'
			) : ''}
```
was enough to start something up.
It seems like all dynamic imports are first loaded relatively (in the else case) and absolutely on subsequent imports and it was importing the case for systemjs so nothing happened.

systemjs in it's current iteration is just a polyfill for module loading (or at least s.js, there's also more complicated version).

**Comment by charlag** [comment]
As long as systemjs is not supported, would do you recommend for the browser? es?

**Comment by PepsRyuu** [comment]
ES I would personally recommend, as it's well supported across browsers.

Notable exception of course is Internet Explorer. Is that the issue you're facing?

**Comment by charlag** [comment]
Thanks! Well we don't develop on it so it's not an issue I hope :laughing: 
Since new node also supports it more or less we can try to use it everywhere for debug builds.

**Comment by PepsRyuu** [comment]
Released support for basic template literals for dynamic imports in ```0.12.3```. Also added an error message for when output format contains an unsupported output format. 

From a preliminary investigation of SystemJS, it would require quite a bit of work to support. Not impossible, just incredibly messy as it uses a very different structure from ESM/CJS. Not sure if it should be added considering its use is primarily for IE11 support.

**Comment by charlag** [comment]
Thanks a lot!
As long as I will be able to get es modules running under node, we should be good so maybe we won't need systemjs.

**Comment by charlag** [comment]
Hi!
I get the same error with esm on node now:

```
Cannot read property '__nollup_dynamic_require_callback' of undefined
    at file:///home/ivk/dev/repositories/tutanota-3/build/Env-[hash].js:7:24
    at file:///home/ivk/dev/repositories/tutanota-3/build/Env-[hash].js:111:15
    at ModuleJob.run (internal/modules/esm/module_job.js:140:23)
    at async Loader.import (internal/modules/esm/loader.js:162:24)
```
(seems like it doesn't get `global?`)

and entry point is like this:

```js
import("../../src/api/Env.js").then((module) => {
	module.bootFinished()
	import('./Suite.js')
})
```

**Comment by PepsRyuu** [comment]
Haven't actually tried the ESM support in Node at all. Usually I would recommending setting the format to ```commonjs``` since that's more well supported in general. Would need to investigate Node ESM and see what global variable is shared across all files if any.

**Comment by charlag** [comment]
Thank you for the quick reply!
I switched to esm from your advice above to support both browsers and node with a single build. I suspect it's something small, I will check if I can figure it out.

**Comment by PepsRyuu** [comment]
Might be a case for ```globalThis```.

**Comment by charlag** [comment]
`(typeof self !== 'undefined' ? self : typeof global !== 'undefined' ? global : this)`
works.
Should I send a PR?

**Comment by PepsRyuu** [comment]
Please do! Won't be able to test it out for a couple of days though. Currently on holidays. :)

**Comment by charlag** [comment]
Sorry to bother you on your holiday! Will send it, it's not pressuring right now.

**Comment by PepsRyuu** [comment]
Merged, will be released in next version ```0.13.3```
